DRAFT: Fixes for issue #13

OrangeLightning219 opened the pull request:
- [x] The interface restriction is not parsing correctly: when parsing test :: (a: $T/interface X) {} the restrictions_interface flag is not set to true.
- [x] When parsing inline assembly the body is empty.
- [x] When using push_contex,defer_pop the block variable shouldn't be set. If there is a block after push_context,defer_pop it should be parsed as a separate node.
- [x] The left side of the = operator in a:, b = 1, 2 and a, b = 1, 2 is parsed as a comma separated expression instead of a compound declaration.
- [x] a :: #run -> string { return "a"; } is not parsing correctly and the expression is set to null.
- [x] no_parameters in using is never set.
- [x] There is no way to tell if a comment is single or multi line.
- [x] The #v2 directive in for is ignored
- [x] There is no way to tell if an if statement had a then keyword used.
- [x] In a := ifx b else c; else is ignored.
- [x] In if true then print("true"); else print("false"); else is ignored.
- [x] Intrinsic names are not parsed when they are in quotes. For example#intrinsic "llvm.debugtrap" in Runtime_Support.jai.
- [x] #elsewhere can have an alias after the library name just like #foreign. For example stdin: *FILE #elsewhere libc "__stdinp"; in POSIX/bindings/macos/arm64/stdio.jai. Currently the parser ignores the alias.
- [x] Postfix casts are parsed as comma separated expressions.
- [x] Function style casts are parsed incorrectly and produce an error: token ',' != required ')'. For example a := cast( *u8, b, trunc, force );
- [x] The separation of notes into name and value doesn't really make sense because the notes can have different structures. For example @Note:value, @Test-Note{value=1}, @Note[0] are all valid notes but won't be parsed correctly.
- [x] Unary operator EXPAND is not used anywhere.
- [x] There is no way to tell if a string literal is a here string or a normal string.

OrangeLightning219 commented:
Ok, most of these are small fixes. The biggest changes are:
- Here strings are now parsed in the same way as other directives
- Changed how compound declarations are parsed to hopefully include all possible combinations

Do you have some way to automatically test the changes or do you just manually test the parser output?

The biggest thing left to do is what I thought would be the easiest - not skipping comments 😄.  It seems like it requires massive changes to how the parser works considering that a comment can be placed basically anywhere. I don't have any ideas how to fix this in a simple way but maybe I'll come up with something tomorrow.

SogoCZE commented:
Thanks! I’ll look at the changes soon. I don’t currently have any automatic testing, sadly. Comments should be handled in the lexer before parsing is done, so it should be quite simple to implement the difference between single-line and multiline comments. But perhaps I’m missing something you discovered recently…

OrangeLightning219 commented:
Yeah just figuring out the difference between single-line and multiline is easy. The problem is when I change this:
```
    if is_comment(iterator) {
        parse_comment(iterator); // Skip comments...
        return .{};
    }
```
into
```
    if is_comment(iterator) {
        return parse_comment(iterator);
    }
```
in `parse_next_token` now all the comments are flowing into the parsing logic and are exposed to the user but now when parsing different constructs we have to check if there is a comment or if there is the thing we actually expected. For example parsing `using/*comment1*/,/*comment2*/no_parameters a;` now becomes: expect_using(), check_for_comment(), expect_comma(), check_for_comment(), expect_modifier() etc. So somehow we have to create a Using node while also informing the user that there where comments inside that using expression.

SogoCZE commented:
I see... We can probably implement special handling for comments into `eat_token`, `is_token`, etc., to be able to pick up comments anywhere.

SogoCZE commented:
I see... We should probably create a separate array where we store all comments along with their locations. This is because we actually don’t want comments in the parser. You can use the comments array later to insert the comments during formatting.

OrangeLightning219 commented:
I was thinking about some separate list with comments too. I'll try doing that and see if that's enough information for formatting.

OrangeLightning219 commented:
Ok, I think what I just pushed should be fine. I added a separate list of comments to the lexer so I can use them but the parser still ignores them. I implemented printing comments in my formatter and it seems to be working fine for most of the cases. In some weird cases (like `test :/*comment*/: ()`) I have to reorder the comments (`test :: /*comment*/()`) but it's good enough for me.

SogoCZE commented:
Nice! Thanks! I’ll check the changes, do some testing, and after that, we can merge it. Are you planning to open-source your formatter? It would be interesting to use it inside Jails for formatting and code printing (AST to code).

OrangeLightning219 commented:
Yes the new formatter will be replacing the one I already have on my github. I'll publish it soon because I'm almost done.

Also I noticed that the parser only parses ints, floats and hex ints (other possible number types are hex floats, scientific floats and binary ints). I'll push a fix for that in a moment.

OrangeLightning219 commented on parser.jai:
```diff
@@ -2961,8 +2962,15 @@ parse_if :: (parser: *Parser, kind: If.If_Kind = .UNKNOWN, compile_time := false
-        _if._then = parse(parser, _if);
+        then_found, then_token := maybe_eat_token(parser.lexer, .KEYWORD_THEN);
+        _if.then_keyword_used = then_found;
+        // NOTE: In ifx we only want to parse _then if the 'then' keyword was actually used. For ifx with implicit 'then' _then will be empty because the condition is also treated as _then.
```
Fixed. I always thought that 'then' is required when not using implicit 'then' 😄

SogoCZE commented:
Can we merge it? I’ve done some testing, and everything looks great!

OrangeLightning219 commented:
Yes you can merge it. I'm basically done for now. I'll be testing my formatter now and if I find anything else I'll make another pull request.
